File: go/0416M-分割等和子集.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-29 10:21:17
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func canPartition(nums []int) bool {
	sum := 0
	for _, v := range nums {
		sum += v
	}
	if sum&1 == 1 { // 奇数不可能
		return false
	}
	sum /= 2
	dp := make([]int, sum+1)
	for i := range nums {
		for j := sum; j >= nums[i]; j-- {
			dp[j] = max(dp[j], dp[j-nums[i]]+nums[i])
		}
	}
	if dp[sum] == sum {
		return true
	}
	return false
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: cpp/0416M-分割等和子集.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-01-10 12:25:59
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum & 1) return false;
        sum >>= 1;  // 填满半个背包
        vector<int> dp(sum + 1);
        for (int i = 0; i < nums.size(); i++) {
            for (int j = sum; j >= nums[i]; j--) {
                dp[j] = max(dp[j], dp[j - nums[i]] + nums[i]);
            }
        }
        if (dp[sum] == sum) return true;
        return false;
    }
};